src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Utilities/ParsedDocument.cs

```diff
@@ -55,4 +58,33 @@ public ParsedDocument WithChangedRoot(SyntaxNode root, CancellationToken cancell
         var text = root.SyntaxTree.GetText(cancellationToken);
         return new ParsedDocument(Id, text, root, HostLanguageServices);
     }
+
+    public ParsedDocument WithChange(TextChange change, CancellationToken cancellationToken)
+        => WithChangedText(Text.WithChanges(change), cancellationToken);
+
+    public ParsedDocument WithChanges(IEnumerable<TextChange> changes, CancellationToken cancellationToken)
+        => WithChangedText(Text.WithChanges(changes), cancellationToken);
+
+    /// <summary>
+    /// Equivalent semantics to <see cref="Document.GetTextChangesAsync(Document, CancellationToken)"/>
+    /// </summary>
+    public IEnumerable<TextChange> GetChanges(in ParsedDocument oldDocument)
+    {
+        Contract.ThrowIfFalse(Id == oldDocument.Id);
+
+        if (Text == oldDocument.Text || SyntaxTree == oldDocument.SyntaxTree)
+        {
+            return SpecializedCollections.EmptyEnumerable<TextChange>();
+        }
+
+        var textChanges = Text.GetTextChanges(oldDocument.Text);
+
+        // if changes are significant (not the whole document being replaced) then use these changes
+        if (textChanges.Count > 1 || textChanges.Count == 1 && textChanges[0].Span != new TextSpan(0, oldDocument.Text.Length))
```

Comment:
i genuinely find this expression too confusing to understand :)  can you break up or use parens?

Reply:
Added line break